### DISCUSSION TITLE
Consultas varias

### DISCUSSION THREAD

**Discussion by Gdepablo** [dds-jv/foro#383] [Consultas de persistencia] [answered]
Buenas @gastonprieto, tenia dudas con la clase del viernes pasado
No entiendo bien primero qué se supone que hace el discriminator (tampoco el discriminator value), y si cumple alguna función útil
Vi que en la clase se habla de agregar, también, discriminador de más o de menos, y eso no termino de entender a qué se referían
Después el tema de JOINED, que vi que se habla de ventajas y desventajas, ¿qué se supone que hace? ¿Te permite unir dos tablas? Pero cómo es que funciona?
Tampoco me queda claro cuando hablás de "pisar el palito" que hay que fijarse en el DER el tema de la estrategia (diste el ejemplo de 10 atributos, y estábamos hablando sobre la lista de jugadores)
Y ¿cómo se gestionan bien los ID? Eso de que las clases heredadas comparten numeración, ¿lo gestiona todo JPA/Hibernate? ¿O cómo va la cosa?

Me pareció escuchar que se mencionaba algo de que las transacciones y los tests lo hacía todo la biblioteca de tests (en referencia a los settings de la DB) pero no entendí, o sea, que en los test no tenemos que poner el @Entity ni nada porque lo hace automáticamente el framework?

EDIT: Cuál es la clase que hay que ver por el feriado? La que dice ver las últimas dos horas?

**Comment by flbulgarelli** [comment, marked as answer]
¡Hola!

> No entiendo bien primero qué se supone que hace el discriminator (tampoco el discriminator value), y si cumple alguna función útil

La columna discriminadora es esencial para poder implementar una estrategia de mapeo de herencia _Single Table_ (STI), porque de lo contrario, no es posible saber a qué clase corresponde una determinada fila. `@DiscriminatorValue` sirve para establecer los posibles valores de ese discriminador, mientras que `@DiscriminatorColumn` establece el nombre de dicha columna. Para más información, te sugiero consultar [la sección 5.3 del apunte,](https://docs.google.com/document/d/1YLmp9vMnSzKg2emt3Bx564Tf1CLalShPc98Z8nCoi7s/edit) la [documentación](https://docs.oracle.com/javaee/7/api/javax/persistence/DiscriminatorColumn.html), [la guía de anotaciones](https://docs.google.com/document/d/1jWtehhVCFYECKvpdcCxnEgWZFCv2fR2WPyUJSoiX3II/edit) y/o volver a ver la clase del año pasado (clase 14, minuto 1:30:00 aprox) o la del viernes pasado (clase 15, acabo de subir su video al campus).


> Después el tema de JOINED, que vi que se habla de ventajas y desventajas, ¿qué se supone que hace? ¿Te permite unir dos tablas? Pero cómo es que funciona?

_Joined_ es otra estrategia de mapeo de herencia, al igual que Single Table o Table Per Class. A diferencia de single table, en la que todas las instancias de una jerarquía de herencia se mapean a filas en una una única tabla, en joined se mapean a filas de múltiples tablas, cada una representando una clase de la jerarquía (ya sea concreta o abstracta).  Te sugiero ver la sección 5.2 del apunte, en la que encontrarás ejemplos de cómo se mapea a tablas, y la sección 5.4, en la que encontrarás una comparativa. 

:warning: No confundir con `@JoinColumn`, que es una anotación que permite establecer la columna que se utilizará para unir dos tablas en una relación entre dos entidades (es decir, una FK). 
 

>  Y ¿cómo se gestionan bien los ID? Eso de que las clases heredadas comparten numeración, ¿lo gestiona todo JPA/Hibernate? ¿O cómo va la cosa? 

Esto es correcto, bajo la estrategia Joined en general la numeración es única: no habrá dos instancias dentro de la misma jerarquía que compartan el mismo ID. Eso es resuelto por el propio framework, que genera los IDs de la PK en la tabla correspondiente a la clase (típicamente abstracta) común, y luego utiliza al mismo valor de ID en las FKs correspondientes de las clases hijas. 


> Me pareció escuchar que se mencionaba algo de que las transacciones y los tests lo hacía todo la biblioteca de tests (en referencia a los settings de la DB) pero no entendí, o sea, que en los test no tenemos que poner el https://github.com/entity ni nada porque lo hace automáticamente el framework?

Acá no te estoy siguiendo. Los tests no son entidades, porque representan objetos persistentes. Luego, si querés que los tests sean transaccionales, JPA/Hibernate tampoco hace nada por vos: sos completamente responsable de abrir y deshacer transacciones en los mismos para que los objetos persistidos en ellos no perduren realmente. Sin embargo, en general los frameworks que se paran encima de los ORMs ofrecen abstracciones similares a las de un test transaccional. Por ejemplo, si usás `jpa-java8-extras` y heredás de `AbstractPersistenceTest` este comportamiento se vuelve automático bajo Junit4 (o Junit5, con ciertos cuidados).

 
> EDIT: Cuál es la clase que hay que ver por el feriado? La que dice ver las últimas dos horas?

Sí te referís al 19/8, sí. Si te referís a 2/9, no, es el video de la clase 14.
